File: objective-c-sample/SendBirdSampleUI/SendBird/SendBirdChat/View/MessageTableViewCell.m
//
//  MessageTableViewCell.m
//  SendBirdSampleUI
//
//  Created by SendBird Developers on 2015. 7. 29..
//  Copyright (c) 2015 SENDBIRD.COM. All rights reserved.
//

#import "MessageTableViewCell.h"

#define kMessageCellTopMargin 4
#define kMessageCellBottomMargin 4
#define kMessageCellLeftMargin 15
#define kMessageCellRightMargin 15
#define kMessageFontSize 14.0

@implementation MessageTableViewCell

- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    
    if (self != nil) {
        [self initViews];
    }
    
    return self;
}

- (void) initViews
{
    [self.contentView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.contentView setBackgroundColor:[UIColor clearColor]];
    [self setBackgroundColor:[UIColor clearColor]];
    
    self.messageLabel = [[UILabel alloc] init];
    [self.messageLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.messageLabel setFont:[UIFont systemFontOfSize:kMessageFontSize]];
    [self.messageLabel setNumberOfLines:0];
    [self.messageLabel setLineBreakMode:NSLineBreakByCharWrapping];
    
    [self.contentView addSubview:self.messageLabel];
    
    // Message Label
    [self.contentView addConstraint:[NSLayoutConstraint constraintWithItem:self.messageLabel
                                                                 attribute:NSLayoutAttributeTop
                                                                 relatedBy:NSLayoutRelationEqual
                                                                    toItem:self.contentView
                                                                 attribute:NSLayoutAttributeTop
                                                                multiplier:1 constant:kMessageCellTopMargin]];
    
    [self.contentView addConstraint:[NSLayoutConstraint constraintWithItem:self.messageLabel
                                                                 attribute:NSLayoutAttributeLeading
                                                                 relatedBy:NSLayoutRelationEqual
                                                                    toItem:self.contentView
                                                                 attribute:NSLayoutAttributeLeading
                                                                multiplier:1 constant:kMessageCellLeftMargin]];
    
    [self.contentView addConstraint:[NSLayoutConstraint constraintWithItem:self.messageLabel
                                                                 attribute:NSLayoutAttributeTrailing
                                                                 relatedBy:NSLayoutRelationEqual
                                                                    toItem:self.contentView
                                                                 attribute:NSLayoutAttributeTrailing
                                                                multiplier:1 constant:-kMessageCellRightMargin]];
    
    [self.contentView addConstraint:[NSLayoutConstraint constraintWithItem:self.messageLabel
                                                                 attribute:NSLayoutAttributeBottom
                                                                 relatedBy:NSLayoutRelationEqual
                                                                    toItem:self.contentView
                                                                 attribute:NSLayoutAttributeBottom
                                                                multiplier:1 constant:kMessageCellBottomMargin]];
    
    // Content View
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.contentView
                                                     attribute:NSLayoutAttributeTop
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTop
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.contentView
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.contentView
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:0]];
}

- (void) setModel:(SendBirdMessage *)message
{
    self.message = message;
    [self.messageLabel setAttributedText:[self buildMessage]];
}

- (NSAttributedString *)buildMessage
{
    NSMutableDictionary *nameAttribute = [NSMutableDictionary dictionaryWithObjectsAndKeys:[UIFont boldSystemFontOfSize:kMessageFontSize], NSFontAttributeName, UIColorFromRGB(0x824096), NSForegroundColorAttributeName, nil];
    NSMutableDictionary *messageAttribute = [NSMutableDictionary dictionaryWithObjectsAndKeys:[UIFont systemFontOfSize:kMessageFontSize], NSFontAttributeName, UIColorFromRGB(0x595959), NSForegroundColorAttributeName,nil];
    NSMutableDictionary *urlAttribute = [NSMutableDictionary dictionaryWithObjectsAndKeys:[UIFont systemFontOfSize:kMessageFontSize], NSFontAttributeName, UIColorFromRGB(0x01579b), NSForegroundColorAttributeName,nil];
    
    NSString *message = [[NSString stringWithFormat:@"%@: %@", [[self.message sender] name], [self.message message]] stringByReplacingOccurrencesOfString:@" " withString:@"\u00A0"];
    NSString *url = [SendBirdUtils getUrlFromString:[self.message message]];
    NSRange urlRange;
    if ([url length] > 0) {
        urlRange = [message rangeOfString:url];
    }
    message = [message stringByReplacingOccurrencesOfString:@" " withString:@"\u00A0"];
    message = [message stringByReplacingOccurrencesOfString:@"-" withString:@"\u2011"];
    
    int badge = 0;
    if ([self.message isOpMessage]) {
        message = [NSString stringWithFormat:@"\u00A0\u00A0%@", message];
        badge = 2;
    }
    
    NSMutableAttributedString *attributedMessage = [[NSMutableAttributedString alloc] initWithString:message];
    NSRange nameRange = NSMakeRange(badge, [[[self.message sender] name] length]);
    NSRange messageRange = NSMakeRange([[[self.message sender] name] length] + badge, [[self.message message] length] + 2);
    
    if ([self.message isOpMessage]) {
        NSTextAttachment *textAttachment = [[NSTextAttachment alloc] init];
        textAttachment.image = [UIImage imageNamed:@"_icon_admin"];
        NSAttributedString *attrStringWithImage = [NSAttributedString attributedStringWithAttachment:textAttachment];
        [attributedMessage replaceCharactersInRange:NSMakeRange(0, 1) withAttributedString:attrStringWithImage];
    }
    
    [attributedMessage beginEditing];
    [attributedMessage setAttributes:nameAttribute range:nameRange];
    [attributedMessage setAttributes:messageAttribute range:messageRange];
    if ([url length] > 0) {
        [attributedMessage setAttributes:urlAttribute range:urlRange];
    }
    [attributedMessage endEditing];
    
    return attributedMessage;
}

- (CGFloat)getHeightOfViewCell:(CGFloat)totalWidth
{
    CGFloat messageWidth;
    CGRect messageRect;
    NSAttributedString *attributedMessage = [self buildMessage];
    
    messageWidth = totalWidth - (kMessageCellLeftMargin + kMessageCellRightMargin);
    messageRect = [attributedMessage boundingRectWithSize:CGSizeMake(messageWidth, CGFLOAT_MAX) options:(NSStringDrawingUsesLineFragmentOrigin) context:nil];
    
    return messageRect.size.height + kMessageCellTopMargin + kMessageCellBottomMargin;
}

@end

File: swift-sample-v2/SendBirdSampleUI/SendBird/SendBirdChat/View/MessageTableViewCell.swift
//
//  MessageTableViewCell.swift
//  SendBirdSampleUI
//
//  Created by Jed Kyung on 2/4/16.
//  Copyright © 2016 SENDBIRD.COM. All rights reserved.
//

import UIKit
import SendBirdSDK

class MessageTableViewCell: UITableViewCell {
    let kMessageCellTopMargin: CGFloat = 4
    let kMessageCellBottomMargin: CGFloat = 4
    let kMessageCellLeftMargin: CGFloat = 15
    let kMessageCellRightMargin: CGFloat = 15
    let kMessageFontSize: CGFloat = 14.0

    var message: SendBirdMessage?
    var messageLabel: UILabel?

    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        self.initViews()
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private func initViews() {
        self.contentView.translatesAutoresizingMaskIntoConstraints = false
        self.contentView.backgroundColor = UIColor.clearColor()
        self.backgroundColor = UIColor.clearColor()

        self.messageLabel = UILabel()
        self.messageLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.messageLabel?.font = UIFont.systemFontOfSize(kMessageFontSize)
        self.messageLabel?.numberOfLines = 0
        self.messageLabel?.lineBreakMode = NSLineBreakMode.ByCharWrapping
        
        self.contentView.addSubview(self.messageLabel!)

        
        // Message Label
        self.contentView.addConstraint(NSLayoutConstraint.init(item: self.messageLabel!, attribute: NSLayoutAttribute.Top, relatedBy: NSLayoutRelation.Equal, toItem: self.contentView, attribute: NSLayoutAttribute.Top, multiplier: 1, constant: kMessageCellTopMargin))
        self.contentView.addConstraint(NSLayoutConstraint.init(item: self.messageLabel!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self.contentView, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: kMessageCellLeftMargin))
        self.contentView.addConstraint(NSLayoutConstraint.init(item: self.messageLabel!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self.contentView, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -kMessageCellRightMargin))
        self.contentView.addConstraint(NSLayoutConstraint.init(item: self.messageLabel!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self.contentView, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: kMessageCellBottomMargin))
        
        // Content View
        self.addConstraint(NSLayoutConstraint.init(item: self.contentView, attribute: NSLayoutAttribute.Top, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Top, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.contentView, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.contentView, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: 0))
    }
    
    func setModel(model: SendBirdMessage) {
        self.message = model;
        self.messageLabel?.attributedText = self.buildMessage()
    }
    
    func buildMessage() -> NSAttributedString {
        let nameAttribute = [NSFontAttributeName: UIFont.boldSystemFontOfSize(kMessageFontSize), NSForegroundColorAttributeName: SendBirdUtils.UIColorFromRGB(0x824096)]
        let messageAttribute = [NSFontAttributeName: UIFont.systemFontOfSize(kMessageFontSize), NSForegroundColorAttributeName: SendBirdUtils.UIColorFromRGB(0x595959)]
        let urlAttribute = [NSFontAttributeName: UIFont.systemFontOfSize(kMessageFontSize), NSForegroundColorAttributeName: SendBirdUtils.UIColorFromRGB(0x01579b)]
        var msg = String.init(format: "%@: %@", (self.message?.sender.name)!, (self.message?.message.stringByReplacingOccurrencesOfString(" ", withString: "\u{00A0}"))!)
        let url: String = SendBirdUtils.getUrlFromstring((self.message?.message)!)
        var urlRange: NSRange?
        if url.characters.count > 0 {
            urlRange = (msg as NSString).rangeOfString(url)
        }
        msg = msg.stringByReplacingOccurrencesOfString(" ", withString: "\u{00A0}")
        msg = msg.stringByReplacingOccurrencesOfString("-", withString: "\u{2011}")
        
        var badge: Int = 0
        if self.message?.isOpMessage == true {
            msg = String.init(format: "\u{00A0}\u{00A0}%@", msg)
            badge = 2
        }
        
        let attributedMessage: NSMutableAttributedString = NSMutableAttributedString.init(string: msg)
        let nameRange: NSRange = NSMakeRange(badge, (self.message?.sender.name.characters.count)!)
        let messageRange: NSRange = NSMakeRange(self.message!.sender.name.characters.count + badge, self.message!.message.characters.count + 2)
        
        if self.message?.isOpMessage == true {
            let textAttachment: NSTextAttachment = NSTextAttachment()
            textAttachment.image = UIImage.init(named: "_icon_admin")
            let attrStringWithImage: NSAttributedString = NSAttributedString.init(attachment: textAttachment)
            attributedMessage.replaceCharactersInRange(NSMakeRange(0, 1), withAttributedString: attrStringWithImage)
        }
        
        attributedMessage.beginEditing()
        attributedMessage.setAttributes(nameAttribute, range: nameRange)
        attributedMessage.setAttributes(messageAttribute, range: messageRange)
        if url.characters.count > 0 {
            attributedMessage.setAttributes(urlAttribute, range: urlRange!)
        }
        attributedMessage.endEditing()
        
        return attributedMessage
    }
    
    func getHeightOfViewCell(totalWidth: CGFloat) -> CGFloat {
        var messageWidth: CGFloat
        var messageRect: CGRect
        let attributedMessage: NSAttributedString = self.buildMessage()
        
        messageWidth = totalWidth - (kMessageCellLeftMargin + kMessageCellRightMargin)
        messageRect = attributedMessage.boundingRectWithSize(CGSizeMake(messageWidth, CGFloat.max), options: NSStringDrawingOptions.UsesLineFragmentOrigin, context: nil)
        
        return messageRect.size.height + kMessageCellTopMargin + kMessageCellBottomMargin
    }
}
